File: mover/jack-751/code/task8/ctf_task8/src/bcs_getproof.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var crypto_1 = require("crypto");
function hexstrToBytes(s) {
    if (s.startsWith('0x')) {
        s = s.slice(2);
    }
    if (s.length % 2 !== 0)
        s = '0' + s;
    var bytes = new Uint8Array(s.length / 2);
    for (var i = 0; i < s.length; i += 2) {
        bytes[i / 2] = parseInt(s.slice(i, i + 2), 16);
    }
    return Buffer.from(bytes);
}
// 參數
var sender_bytes = hexstrToBytes('0xb961b905ab4a1d6ea41e21b29bacff75efc09759bf68decb4733a772f32820b3');
var challenge_bcs = hexstrToBytes('0x19e76ca504c5a5fa5e214a45fca6c058171ba333f6da897b82731094504d5ab90a6d612d6b7a3a3b25596e03000000000000009300000000000000');
var github_id = Buffer.from('jack-751'); // 沒用到
var found = false;
for (var try_val = 100; try_val < 4294967295; try_val++) {
    var proof = Buffer.alloc(4);
    proof.writeUInt32LE(try_val, 0);
    var full_proof = Buffer.concat([proof, sender_bytes, challenge_bcs]);
    var hash_bytes = (0, crypto_1.createHash)('sha3-256').update(full_proof).digest();
    var prefix_sum = 0;
    for (var i = 0; i < 3; i++) {
        prefix_sum += hash_bytes[i];
    }
    if (prefix_sum === 0) {
        found = true;
        console.log('FOUND proof int:', try_val);
        console.log('FOUND proof bytes:', proof);
        console.log('FOUND hash:', hash_bytes.toString('hex'));
        console.log('proof (hex):', proof.toString('hex'));
        break;
    }
}
if (!found) {
    console.log('No valid proof found');
}

File: mover/jack-751/code/task8/ctf_task8/src/bcs_getproof.py
import hashlib
from binascii import unhexlify

def hexstr_to_bytes(s):
    if s.startswith('0x'):
        s = s[2:]
    return unhexlify(s)

# 參數
sender_bytes = hexstr_to_bytes('0xb961b905ab4a1d6ea41e21b29bacff75efc09759bf68decb4733a772f32820b3')
challenge_bcs = hexstr_to_bytes('0x19e76ca504c5a5fa5e214a45fca6c058171ba333f6da897b82731094504d5ab90a6d612d6b7a3a3b25596e03000000000000009300000000000000')

github_id = b'jack-751'  

found = False
for try_val in range(100, 4294967295):
    print('FOUND proof int:', try_val)
    proof = try_val.to_bytes((try_val.bit_length() + 7) // 8 or 1, 'big')
    # print('FOUND proof int:', try_val)
    # print('proof (bytes):', proof)
    # print('proof (hex):', proof.hex())
    # print('sender_bytes:', sender_bytes)
    # print('challenge_bcs:', challenge_bcs)
    full_proof = proof + sender_bytes + challenge_bcs
    # print('proof (bytes):', proof)
    # print('proof (hex):', proof.hex())
    # print('sender_bytes:', sender_bytes.hex())
    # print('challenge_bcs:', challenge_bcs.hex())
    # print('full_proof:', full_proof.hex())
    hash_bytes = hashlib.sha3_256(full_proof).digest()
    print('hash:', hash_bytes.hex())

    prefix_sum = 0
    for i in range(3):
        a = hash_bytes[i]
        # print('a:', a)
        # print('prefix_sum:', prefix_sum)
        # print('----------------')
        prefix_sum += a
        # print('proof:', proof)
        # print('hash:', hash_bytes.hex())

    if prefix_sum == 0:
        found = True
        print('FOUND proof int:', try_val)
        print('FOUND proof bytes:', proof)
        print('FOUND hash:', hash_bytes.hex())
        print('proof (hex):', proof.hex())
        break

if not found:
    print('No valid proof found')